graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/dml/MutationFetcher.java

```diff
@@ -20,11 +41,75 @@ protected MutationFetcher(
   }
 
   @Override
-  protected Map<String, Object> get(DataFetchingEnvironment environment, DataStore dataStore)
-      throws Exception {
-    String statement = buildStatement(environment, dataStore);
-    dataStore.query(statement).get();
-    return ImmutableMap.of("value", environment.getArgument("value"));
+  protected CompletableFuture<Map<String, Object>> get(
+      DataFetchingEnvironment environment, DataStore dataStore) {
+    String statement = null;
+    Exception buildException = null;
+
+    // Avoid mixing sync and async exceptions
+    try {
+      // buildStatement() could throw an unchecked exception.
+      // As the statement might be part of a batch, we need to make sure the
+      // batched operation completes.
+      statement = buildStatement(environment, dataStore);
+    } catch (Exception e) {
+      buildException = e;
+    }
+
+    OperationDefinition operation = environment.getOperationDefinition();
+
+    if (operation.getDirectives().stream().anyMatch(d -> d.getName().equals(ATOMIC_DIRECTIVE))
+        && operation.getSelectionSet().getSelections().size() > 1) {
+      // There are more than one mutation in @atomic operation
+      return executeAsBatch(environment, dataStore, statement, buildException);
+    }
+
+    if (buildException != null) {
+      CompletableFuture<Map<String, Object>> f = new CompletableFuture<>();
+      f.completeExceptionally(buildException);
+      return f;
+    }
+
+    // Execute as a single statement
+    return dataStore
+        .query(statement)
+        .thenApply(rs -> ImmutableMap.of("value", environment.getArgument("value")));
+  }
+
+  private CompletableFuture<Map<String, Object>> executeAsBatch(
+      DataFetchingEnvironment environment,
+      DataStore dataStore,
+      String statement,
+      Exception buildException) {
+    int selections = environment.getOperationDefinition().getSelectionSet().getSelections().size();
+    HTTPAwareContextImpl context = environment.getContext();
+    HTTPAwareContextImpl.BatchContext batchContext = context.getBatchContext();
+
+    if (environment.getArgument("options") != null) {
+      // Users should specify query options once in the batch
+      int counter = batchContext.setDataStore(dataStore);
```

Comment:
The counter value itself isn't ever used. Maybe it would be clearer if this returned boolean value or threw a checked exception: "Datastore already set".

Reply:
Good point, I'll fix it.

Reply:
Just pushed a fix.